drivers/usb/udc/udc_it82xx2.c

```diff
@@ -0,0 +1,1434 @@
+ * EP mapping because when (ep_idx % SHARED_FIFO_NUM) is 3, it actually means the EP0.
+ */
+static const uint8_t ep_fifo_res[SHARED_FIFO_NUM] = {3, 1, 2};
+
+static int it82xx2_usb_fifo_ctrl(const struct device *dev, uint8_t ep)
+{
+	const uint8_t ep_idx = USB_EP_GET_IDX(ep);
+	const struct usb_it82xx2_config *config = dev->config;
+	struct usb_it82xx2_regs *const usb_regs = config->base;
+	volatile uint8_t *ep_fifo_ctrl = usb_regs->fifo_regs[EP_EXT_REGS_BX].fifo_ctrl.ep_fifo_ctrl;
+	uint8_t fifon_ctrl = (ep_fifo_res[ep_idx % SHARED_FIFO_NUM] - 1) * 2;
+	int ret = 0;
+
+	if (ep_idx == 0) {
+		LOG_ERR("Invalid endpoint 0x%x", ep);
+		return -EINVAL;
+	}
+
+	if (USB_EP_DIR_IS_IN(ep)) {
+		if (ep_idx < 8) {
+			ep_fifo_ctrl[fifon_ctrl] = BIT(ep_idx);
+			ep_fifo_ctrl[fifon_ctrl + 1] = 0x0;
+		} else {
+			ep_fifo_ctrl[fifon_ctrl] = 0x0;
+			ep_fifo_ctrl[fifon_ctrl + 1] = BIT(ep_idx - 8);
+		}
+	} else if (USB_EP_DIR_IS_OUT(ep)) {
+		if (ep_idx < 8) {
+			ep_fifo_ctrl[fifon_ctrl] |= BIT(ep_idx);
+		} else {
+			ep_fifo_ctrl[fifon_ctrl + 1] |= BIT(ep_idx - 8);
+		}
+	} else {
+		LOG_ERR("Failed to set fifo control register for ep 0x%x", ep);
+		ret = -EINVAL;
+	}
+
+	return ret;
+}
+
+static void it82xx2_event_submit(const struct device *dev, const uint8_t ep,
+				 const enum it82xx2_event_type event)
+{
+	struct it82xx2_ep_event evt;
+
+	evt.dev = dev;
+	evt.ep = ep;
+	evt.event = event;
+	k_msgq_put(&evt_msgq, &evt, K_NO_WAIT);
+}
+
+static int it82xx2_ep_enqueue(const struct device *dev, struct udc_ep_config *const cfg,
+			      struct net_buf *const buf)
+{
+	udc_buf_put(cfg, buf);
+
+	it82xx2_event_submit(dev, cfg->addr, IT82xx2_EVT_XFER);
+	return 0;
+}
+
+static int it82xx2_ep_dequeue(const struct device *dev, struct udc_ep_config *const cfg)
+{
+	struct net_buf *buf;
+
+	buf = udc_buf_get_all(dev, cfg->addr);
+	if (buf) {
+		udc_submit_ep_event(dev, buf, -ECONNABORTED);
+	}
+
+	udc_ep_set_busy(dev, cfg->addr, false);
+
+	return 0;
+}
+
+static int it82xx2_ep_set_halt(const struct device *dev, struct udc_ep_config *const cfg)
+{
+	const uint8_t ep_idx = USB_EP_GET_IDX(cfg->addr);
+	const struct usb_it82xx2_config *config = dev->config;
+	struct usb_it82xx2_regs *const usb_regs = config->base;
+	struct it82xx2_usb_ep_regs *ep_regs = usb_regs->usb_ep_regs;
+	struct gctrl_it8xxx2_regs *const gctrl_regs = GCTRL_IT8XXX2_REGS_BASE;
+	struct it82xx2_priv_data *priv = udc_get_private(dev);
+
+	if (ep_idx == 0) {
+		unsigned int lock_key;
+		uint32_t idx = 0;
+
+		priv->stall_is_sent = true;
+		lock_key = irq_lock();
+		it82xx2_usb_set_ep_ctrl(dev, ep_idx, EP_STALL_SEND, true);
+		it82xx2_usb_set_ep_ctrl(dev, ep_idx, EP_READY_ENABLE, true);
+		/* polling if stall send for 3ms */
+		while (idx < 198 && !(ep_regs[ep_idx].ep_status & DC_STALL_SENT)) {
+			/* wait 15.15us */
+			gctrl_regs->GCTRL_WNCKR = 0;
+			idx++;
+		}
+
+		if (idx < 198) {
+			it82xx2_usb_set_ep_ctrl(dev, ep_idx, EP_STALL_SEND, false);
+		}
```

Comment:
Why do you need to clear this flag? It should actually be done by the controller when new setup packed is received.

Reply:
The bit we clears is to send STALL in response to the controller initiated transaction, i.e., the chip responses STALL if this bit is not cleared.